### PR TITLE
Revise the "validating linear texture data" algorithm

### LINKED ISSUE #984: bytesPerRow and rowsPerImage disagree on validation between 0 and min
We have two validation rules on rowsPerImage:

> - If layout.rowsPerImage is not 0, it must be greater than or equal to copyExtent.height.
> - If copyExtent.depth is greater than 1:
>     layout.rowsPerImage must be greater than or equal to copyExtent.height.

But just one on bytesPerRow:

> - If copyExtent.height is greater than 1 or copyExtent.depth is greater than 1:
>     layout.bytesPerRow must be greater than or equal to the number of bytesInACompleteRow.

These disagree on whether `0` is special; that is:
- If depth=1, rowsPerImage allows only 0 or &ge; height.
- But if height=1 and depth=1, bytesPerRow allows any value even if &lt; bytesInACompleteRow

These should be consistent.

### LINKED ISSUE #987: bytesPerRow should be validated only if copyExtent.height is greater than one *block* height
> For other members in layout:
```diff
-- If copyExtent.height is greater than 1 or copyExtent.depth is greater than 1:
+- If copyExtent.height is greater than 1 block height or copyExtent.depth is greater than 1:
     layout.bytesPerRow must be greater than or equal to the number of bytesInACompleteRow.
 - If copyExtent.depth is greater than 1:
     layout.rowsPerImage must be greater than or equal to copyExtent.height.
```

[EDIT: clarified]

### PR DESCRIPTION
**Pull request by kainino0x**
- Reformats this algorithm to have explicit steps
- Removes the code block (to remove ambiguities about integer overflows, etc.)
- Removes a redundant rule about overflowing a GPUSize64
- Fixes #987: bytesPerRow only required if height \*in blocks\* > 1
- ~Fixes `#984`: rowsPerImage should not be validated unless used~

***
<a href="https://pr-preview.s3.amazonaws.com/kainino0x/gpuweb/pull/1006.html" title="Last updated on Aug 19, 2020, 6:47 PM UTC (35f7faf)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/1006/8a008ad...kainino0x:35f7faf.html" title="Last updated on Aug 19, 2020, 6:47 PM UTC (35f7faf)">Diff</a>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -4215,48 +4215,58 @@ and {{GPUCommandEncoder/copyTextureToBuffer()}}.
...
+    1. It is assumed that |copyExtent|.[=Extent3D/width=] is a multiple of |blockWidth|
+        and |copyExtent|.[=Extent3D/height=] is a multiple of |blockHeight|. Let:
+            - |widthInBlocks| be |copyExtent|.[=Extent3D/width=] &divide; |blockWidth|.
+            - |heightInBlocks| be |copyExtent|.[=Extent3D/height=] &divide; |blockHeight|.
+            - |bytesInACompleteRow| be |blockSize| &times; |widthInBlocks|.
+
+    1. Fail if the following conditions are not satisfied:
+        <div class=validusage>
+            - If |heightInBlocks| &gt; 1 or |copyExtent|.[=Extent3D/depth=] &gt; 1:
+                - |layout|.{{GPUTextureDataLayout/bytesPerRow}} must be
```
Comment (kvark) [inline spec/index.bs:4244]: It seems a bit weird if you could put *anything* in this field if it's not required. Maybe we should at least make sure that it covers the data portion (i.e. `|widthInBlocks| * |blockSize|`

File: spec/index.bs [hunk 2]
```diff
@@ -4215,48 +4215,58 @@ and {{GPUCommandEncoder/copyTextureToBuffer()}}.
...
+                    greater than or equal to the number of |bytesInACompleteRow|.
+            - If |copyExtent|.[=Extent3D/depth=] &gt; 1:
+                - |layout|.{{GPUTextureDataLayout/rowsPerImage}} must not be 0.
+
+            These ensure the copy's subregions of the linear data do not overlap.
+        </div>
+
+    1. If |widthInBlocks|, |heightInBlocks|, or |copyExtent|.[=Extent3D/depth=] is 0:
```
Comment (kvark) [inline spec/index.bs:4252]: speaking of the Myles question on what you could do with zero-sized buffer: you can issue copy operations on it :)

File: spec/index.bs (hunk 1 above)
Comment (kainino0x) [reply spec/index.bs:4244]: That's exactly what bytesInACompleteRow is, but we didn't want to require specifying bytesPerRow for every copy (hence originally allowing 0 in the first place). If we want to disallow arbitrary numbers here should we make bytesPerRow and rowsPerImage optional?

Comment (kvark) [reply spec/index.bs:4244]: yeah, that's in line with the previous out-of-band-messaging changes we've been doing.

File: spec/index.bs [hunk 3]
```diff
@@ -4215,48 +4215,58 @@ and {{GPUCommandEncoder/copyTextureToBuffer()}}.
 
 Issue: define this as an algorithm with (texture, mipmapLevel) parameters and use the call syntax instead of referring to the definition by label.
 
-<div algorithm class=validusage>
+<div algorithm>
```
Comment (toji) [inline spec/index.bs:4218]: The spec flips back and forth between declaring algorithms as functions like this and declaring them with prose (like https://gpuweb.github.io/gpuweb/#abstract-opdef-the-steps-to-serialize-a-gpudevice-object). Feels like we should probably pick one style and stick with it?

FWIW I don't have much in the way of opinons about which to prefer. For more complex algorithms this feels a bit cleaner, though possibly easy to confuse as an API-facing method at first glance. For smaller algorithms it feels overly bulky.

Comment (kainino0x) [reply spec/index.bs:4218]: I think it's probably okay to have a mix of both depending on what fits the case best. I don't think confusing with API facing methods will be too much of a problem: those are nested in `<dl>` and use `argumentdef`. But they have gotten fairly similar.

In the case of the serialization algorithms in particular, we don't define the types of the arguments currently (they're implicitly the same as in the HTML spec link right next to them) but we could use this newer format.

**Comment by kainino0x** [comment]
Unfixed #984 to defer it to another PR. Also made a fix to a piece of validation that I think was wrong from #958? @kvark PTAQL

**Pull request merged**